crescent/plugin.py

```diff
@@ -44,13 +54,24 @@ def load(self, path: str, refresh: bool = False) -> Plugin:
 
         Args:
             path: The module path for the plugin.
-            refresh: Whether or not to reload the plugin.
+            refresh: Whether or not to reload the plugin's module.
         """
 
         plugin = Plugin._from_module(path, refresh=refresh)
-        self.add_plugin(plugin, force=refresh)
+        self._add_plugin(plugin, force=refresh)
+
+        for callback in plugin._load_hooks:
+            callback()
+
         return plugin
 
+    def _add_plugin(self, plugin: Plugin, force: bool = False) -> None:
+        if plugin.name in self.plugins:
+            if not force:
+                raise ValueError(f"Plugin name {plugin.name} already exists.")
+        self.plugins[plugin.name] = plugin
```

Comment:
self.plugins[plugin.name] references the original plugin under the first if-branch. You can call unload hooks there. In fact, you should call `self.unload(name)` if it already exists otherwise you'll still have the duplicate event subscription problem

Reply:
I think the best thing to do right now is a warning then later on make it an Exception. Calling `self.unload` implicitly feels weird to me because the function name `load` says nothing about unloading or reloading.

Reply:
It only calls `self.unload` if the user passes `bot.plugins.load(..., refresh=True)`. To me, it's worse that the `unload` callback for a Plugin isn't called when the plugin is removed, unless a user explicitly calls `bot.plugins.unload` first.

Also, there's still the matter of subscribing events twice. Maybe we deprecate the `refresh=True` part of plugins, but I think it still needs to unload the plugin if it is passed.

Reply:
I think we disagree on the semantics of the load function. Making it automatically reload would be a good idea but I'll need to change the description to make that clear.
I don't think `refresh` should be deprecated. If anything it should default to true. I don't see a reason to prevent users from choosing if a file is re-imported or not.

Reply:
Right now, here's what happens:
```py
bot.plugins.load("path")  # loads the plugin
bot.plugins.load("path")  # raises an error, plugin already exists!
bot.plugins.load("path", refresh=True)  # loads the plugin by replacing the old one, but doesn't call the unload hooks on the currently loaded plugin
bot.plugins.unload("plugin-name")  # removes the plugin and calls unload hooks
```

I think that if refresh=True is passed, that's basically the equivalent of `plugins.reload`. Right now all it does is cause two of the same plugins to exist at the same time - it's not "refreshing", it's "duplicating".

Reply:
I see.

Reply:
Ultimately it's up to you, but at the very least I think that events being duplicated can't be allowed, or it needs to be documented under `refresh`. Though I can't really imagine anyone wanting the plugin to be duplicated and not have the original unloaded.

Reply:
This would be another PR probably, but maybe we should remove/deprecate the refresh parameter and use these methods:
```py
plugins.load  # loads a plugin, raising an error if it already exists
plugins.unload  # unloads a plugin, raising an error if it didn't exist
plugins.reload  # calls plugin.unload and then plugins.load
```

Reply:
This isn't possible without storing a map of plugin paths to names

Reply:
Nvm, had a breakthough.
Now
```
plugins.load  # loads
plugins.load(..., refresh=True)  # reloads
plugins.unload  # unloads
```

Reply:
> This isn't possible without storing a map of plugin paths to names. Ig I gotta find a clean way to do that.

Not true. You can call Plugin._from_path and get its name.

Reply:
Nevermind, I see that you did this lol.